source/extensions/filters/http/decompressor/decompressor_filter.cc

```diff
@@ -72,7 +73,33 @@ Http::FilterHeadersStatus DecompressorFilter::decodeHeaders(Http::RequestHeaderM
   //      the upstream that this hop is able to decompress responses via the Accept-Encoding header.
   if (config_->responseDirectionConfig().decompressionEnabled() &&
       config_->requestDirectionConfig().advertiseAcceptEncoding()) {
-    headers.appendInline(accept_encoding_handle.handle(), config_->contentEncoding(), ",");
+    if (Runtime::runtimeFeatureEnabled(
+            "envoy.reloadable_features.append_to_accept_content_encoding_only_once")) {
+      // Append the content encoding only if it isn't already present in the
+      // accept_encoding header. If it is present with a q-value ("gzip;q=0.3"),
+      // remove the q-value to indicate that the content encoding setting that we
+      // add has max priority (i.e. q-value 1.0).
+      std::vector<absl::string_view> newContentEncodings;
+      std::vector<absl::string_view> contentEncodings =
+          Http::HeaderUtility::parseCommaDelimitedHeader(
+              headers.getInlineValue(accept_encoding_handle.handle()));
+      for (absl::string_view contentEncoding : contentEncodings) {
+        absl::string_view strippedEncoding =
+            Http::HeaderUtility::getSemicolonDelimitedAttribute(contentEncoding);
+        if (strippedEncoding != config_->contentEncoding()) {
+          // Add back all content encodings back except for the content encoding that we want to
+          // add. For example, if content encoding is "gzip", this filters out encodings "gzip" and
+          // "gzip;q=0.6".
```

Comment:
what's the desired behavior if you have "gzip" and want to add "gzip;q=0.6" and vice versa?

I see this discussed in comment but I didn't see a test for this; I see the "mock ;q=0" below but I'm not sure if that covers both situations.

Reply:
That's a good question! It's clear that we (in a sense) lose information because we filter out existing encoding strings. However, the spec says that the values are placed in an order of preference based on the q-value, and a missing q-value implies `;q=1`. So we are in effect adding `gzip;q=1`, meaning that's put to the front of the `accept-encoding` preference list anyway, and so `gzip;q=0.6` would not be considered even if it was there. Now with this patch filtering out duplicates and smaller q-values, it just becomes explicit instead of implicit.

So, the desired behavior is never to add `gzip;q=0.6`, and remove `gzip;q=0.6` if it was there when we are adding `gzip`.

Reply:
So I think we need 4 more testcases, including:
 * add "gzip" to a header that already has "gzip;q=0.6". Result should just be "gzip"
 * add "gzip;q=0.6" to a header with "gzip". Result should just be gzip.
 * add "gzip;q=0.7" to a header with "gzip;q=0.6". Result should be "gzip;q=0.7"
 * add "gzip;q=0.6" to a header with "gzip;q=0.7". Result should be "gzip;q=0.7"

WDYT?

Reply:
Ok, I understand now what you mean: if the decompressor factory gives back an encoding with a q-value, the decompressor filter should figure out what is the greatest common q-value and use that. So far the code doesn't handle max(q-value) calculation; instead the code just replaces the existing values with what it gets from the DecompressorFactory as the desired content encoding.

We could implement the max(q-value) algorithm but I'm not sure if that's worth the added complexity, since at the moment neither of the existing DecompressorFactories (gzip and brotli) uses q-values, and it's bit difficult to think of a use case where a less-than-one q-value set by Envoy decompressor would be helpful. @junr03 you have a TODO comment at the `DecompressorFactory::contentEncoding()` -- do you have a plan to take the API forward to some direction that would influence this?

Reply:
I'm ok with that decision but suggest you add the above test cases anyway, verifying the current (wrong) answer, and put comments in code and tests that we have decided to allow the wrong answer for the corner case because at the moment we don't think that will arise under expected use. If usage of the new API does expand in the future, and requires proper merging of q-values then someone will see a warning in the API doc.

Reply:
I found out while writing the tests that the current decompressor filter implementation doesn't allow decompressors with q-values, even in principle. If the decompressor would use a q-value, it would never match with the content-type header of the incoming compressed data, and content-type headers can't have q-values. The matching call is here:

https://github.com/envoyproxy/envoy/blob/main/source/extensions/filters/http/decompressor/decompressor_filter.h#L173

The correct solution would maybe be separating the content-type and q-value from each other in the decompressor config. Then generating the accept-encoding header and comparing content-encoding could be done using the correct config values (`compression-encoding` and `q`, for example). I think that would be an Decompressor API change, and maybe something which could be done in a future PR.

I added one more specific test for removing the q-value.